### PR TITLE
Include ActiveSupport::Testing::TaggedLogging for Rails >= 7

### LINKED ISSUE #2545: undefined local variable or method `tagged_logger' during perform_enqueued_jobs
## What Ruby, Rails and RSpec versions are you using?

Ruby version: 3.0.3p157 (2021-11-24 revision 3fb7d2cadc) [x86_64-darwin21]
Rails version: 7.0.0
RSpec 3.10
  - rspec-core 3.10.1
  - rspec-expectations 3.10.1
  - rspec-mocks 3.10.2
  - rspec-rails 5.0.2
  - rspec-support 3.10.3

## Observed behaviour

Inside an rspec support module, my app is calling `perform_enqueued_jobs`. The spec in question that is using this support module is expecting this call to raise an exception.

When I attempt to run this spec, the following is raised:
```
got #<NameError: undefined local variable or method `tagged_logger'
```

When I debug into `perform_enqueued_jobs`, I am taken into `_assert_nothing_raised_or_warn` inside `activesupport-7.0.0/lib/active_support/testing/assertions.rb`, When debugging the block below, it shows `e` as having the expected exception.

```ruby
        rescue Minitest::UnexpectedError => e
          if tagged_logger && tagged_logger.warn?
            warning = <<~MSG
              #{self.class} - #{name}: #{e.error.class} raised.
              If you expected this exception, use `assert_raises` as near to the code that raises as possible.
              Other block based assertions (e.g. `#{assertion}`) can be used, as long as `assert_raises` is inside their block.
            MSG
            tagged_logger.warn warning
          end

          raise
        end
```
Debugging further reveals that the `if tagged_logger` check above is causing the undefined error.

I reported this on Rails forum [here](https://discuss.rubyonrails.org/t/undefined-local-variable-or-method-tagged-logger-inside-rails-7-activesupport-assertions/79541) and a member of the core team suggested the following:

"This seems to be a bug in RSpec that doesn’t implement the tagged_logger. Please report the issue in the RSpec issue tracker."

## Expected behaviour

The spec should finish and allow `to raise_error` to check for the expected exception. This spec is working as expected in Rails 6.

### PR DESCRIPTION
**Pull request by shanecav84**
Fixes #2545. Not sure if anyone got to this yet. There was also another problem mentioned in that issue about a missing 'name' method, which this PR does not address.

### CODE REVIEW

**Review by pirj — APPROVED** [review]
Thank you.
Do you think it's possible to add a snippet that would fail without this fix? [An example snippet](https://github.com/rspec/rspec-rails/blob/b4f58f44170526edd1f3d25bbb918e2fcdaae7a8/snippets/avoid_fixture_name_collision.rb#L48).

**Comment by shanecav84** [comment]
I've added the snippet. Not quite sure if I've done it correctly. The snippet will fail without the fix.

File: snippets/include_activesupport_testing_tagged_logger.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+require "rspec/rails"
+
+ActiveJob::Base.queue_adapter = :test
+
+# This connection will do for database-independent bug reports
+ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
+
+RSpec.configure do |config|
+  config.use_transactional_fixtures = true
+end
```
Comment (pirj) [inline snippets/include_activesupport_testing_tagged_logger.rb:49]: This seems redundant.

**Comment by pirj** [comment]
Perfect, thank you!

**Comment by pirj** [comment]
I've re-triggered the build as Rails seems to have finally released 5.2.x with a fix for Ruby 2.2 support.

But they also released 6.1.5 and broke something for us.

Do you think [this](https://github.com/rspec/rspec-rails/runs/5508066551?check_suite_focus=true) is related?
```
[ActiveJob] Failed enqueuing TestJob to Test(default): TestError (TestError)
F

Failures:

  1) Foo error raised in perform_enqueued_jobs with block raises the explicitly thrown error
     Failure/Error:
       expect { perform_enqueued_jobs { TestJob.perform_later } }
         .to raise_error(expected_error)

       expected TestError, got #<Minitest::UnexpectedError: Unexpected exception> with backtrace:
         # include_activesupport_testing_tagged_logger.rb:71:in `block (5 levels) in <main>'
         # include_activesupport_testing_tagged_logger.rb:71:in `block (4 levels) in <main>'
         # include_activesupport_testing_tagged_logger.rb:71:in `block (3 levels) in <main>'
     # include_activesupport_testing_tagged_logger.rb:71:in `block (3 levels) in <main>'
```

**Comment by shanecav84** [comment]
Not sure what's up with that failing jRuby test.

**Comment by pirj** [comment]
Quite an interesting case.
@headius Would you like to have a look?

**Comment by JonRowe** [comment]
I tried to install JRuby this morning to take a look at this failure, but it seems broken on M1 macs via asdf at the moment, do we need this to be a snippet? Its currently a regression test as well, what actually does the tagged logger do? Could we write an integration test instead?

**Comment by headius** [comment]
@JonRowe Latest JRuby 9.3 should be working well on M1. Perhaps asdf is not updated to the latest release? (9.3.3).

The failure error is very peculiar. I don't know why `TestError` would not have the `backtrace` method under any circumstance. I do notice it is running 9.3.1.0, which is a couple releases old now.

I'll try to reproduce locally.

**Comment by headius** [comment]
I do see now that it appears to be trying to call `backtrace` against the `TestError` **class**, not against an instance. Unsure why that is happening so far.

**Comment by headius** [comment]
I have not been able to reproduce so far, but I got some other weird errors that seem to be an issue in my environment.

I would suggest trying to upgrade to 9.3 in this PR since it should be upgraded anyway and might fix the issue.

**Comment by shanecav84** [comment]
Upgrading JRuby did not fix the issue. See also https://github.com/rspec/rspec-rails/pull/2580/commits/5d9d3b0d7edbd627510ebb0d0de9b035fd024238 where I instantiated the expected error to get around the JRuby error but that caused other errors.

**Comment by headius** [comment]
I have cleaned up my local env issue and can now reproduce the same error!

**Comment by headius** [comment]
I will try to look into this for a fix for JRuby 9.3.4 but no promises... we are hoping to release next week. If anyone can narrow this down to some specific test that fails, it would be helpful. It almost looks like it is happening *after* the suite completes, but I am unfamiliar with your suites' output.

**Comment by shanecav84** [comment]
@headius 

1. Checkout the branch for this PR
2. Run bundle
3. `cd snippets; ruby include_activesupport_testing_tagged_logger.rb` will run the test.

**Comment by pirj** [comment]
Superseded by #2587 
Thank you for the hard work!

**Pull request closed without merging**